File: .skills/openclaw-skills/skills/daguniang/stock-tools/scripts/fetch_quote.js
#!/usr/bin/env node
const https = require('https');

function inferMarket(code) {
  if (!/^\d{6}$/.test(code)) return null;
  if (code.startsWith('6') || code.startsWith('9') || code.startsWith('5')) return 'sh';
  return 'sz';
}

function fetchText(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 OpenClaw stock-tools',
        'Referer': 'https://finance.sina.com.cn/'
      }
    }, (res) => {
      const chunks = [];
      res.on('data', chunk => chunks.push(chunk));
      res.on('end', () => {
        try {
          const buffer = Buffer.concat(chunks);
          const text = new TextDecoder('gbk').decode(buffer);
          resolve(text);
        } catch (err) {
          reject(err);
        }
      });
    });
    req.on('error', reject);
    req.setTimeout(10000, () => req.destroy(new Error('Request timeout')));
  });
}

function isInvalidPrice({ price, prevClose, open, high, low, volume, amount }) {
  if (!Number.isFinite(price) || price < 0) return true;
  if (price === 0 && prevClose > 0) return true;
  if (price === 0 && open === 0 && high === 0 && low === 0 && volume === 0 && amount === 0) return true;
  return false;
}

function parseSina(code, text) {
  const market = inferMarket(code);
  const symbol = `${market}${code}`;
  const re = new RegExp(`var hq_str_${symbol}="([^"]*)"`);
  const m = text.match(re);
  if (!m) throw new Error(`Quote not found for ${symbol}`);
  const raw = m[1];
  const parts = raw.split(',');
  if (parts.length < 32) throw new Error('Unexpected quote payload');

  const name = parts[0] || code;
  const open = Number(parts[1]);
  const prevClose = Number(parts[2]);
  const price = Number(parts[3]);
  const high = Number(parts[4]);
  const low = Number(parts[5]);
  const volume = Number(parts[8]);
  const amount = Number(parts[9]);
  const date = parts[30] || null;
  const time = parts[31] || null;
  const invalid = isInvalidPrice({ price, prevClose, open, high, low, volume, amount });
  const safePrice = invalid ? null : price;
  const change = !invalid && Number.isFinite(prevClose) ? Number((price - prevClose).toFixed(2)) : null;
  const pct = !invalid && prevClose ? Number((((price - prevClose) / prevClose) * 100).toFixed(2)) : null;

  return {
    code,
    market,
    symbol,
    name,
    open,
    prevClose,
    price: safePrice,
    high,
    low,
    volume,
    amount,
    change,
    pct,
    quoteDate: date,
    quoteTime: time,
    detailUrl: `https://stockpage.10jqka.com.cn/${code}/`,
    valid: !invalid,
    note: invalid ? '报价异常或暂不可用' : null
  };
}

function formatLine(item) {
  const ts = item.quoteDate && item.quoteTime ? `${item.quoteDate} ${item.quoteTime}` : '时间未确认';

  // 非交易时段/未产生有效现价：给更人话的提示，并尽量展示昨收
  if (!item.valid) {
    const prev = Number.isFinite(item.prevClose) && item.prevClose > 0 ? `昨收：${item.prevClose} 元` : '昨收不可用';
    return `${item.code} ${item.name}｜非交易时段或无有效实时报价｜${prev}｜${ts}`;
  }

  const changeText = item.change === null || item.pct === null
    ? '涨跌未确认'
    : `${item.change >= 0 ? '+' : ''}${item.change}（${item.pct >= 0 ? '+' : ''}${item.pct}%）`;
  return `${item.code} ${item.name}｜${item.price} 元｜${changeText}｜${ts}`;
}

async function main() {
  const argv = process.argv.slice(2).filter(Boolean);
  const plain = argv.filter(arg => arg !== '--json');
  const wantJson = argv.includes('--json');
  if (!plain.length) {
    console.error('Usage: fetch_quote.js [--json] <code> [code2 ...]');
    process.exit(1);
  }

  const normalized = plain.map(code => code.trim()).filter(code => /^\d{6}$/.test(code));
  if (!normalized.length) {
    console.error('No valid 6-digit stock codes provided');
    process.exit(1);
  }

  const symbols = normalized.map(code => `${inferMarket(code)}${code}`).join(',');
  const text = await fetchText(`https://hq.sinajs.cn/list=${symbols}`);
  const result = normalized.map(code => parseSina(code, text));

  if (wantJson) {
    process.stdout.write(JSON.stringify(result, null, 2));
    return;
  }

  process.stdout.write(result.map(formatLine).join('\n'));
}

main().catch(err => {
  console.error(err.message || String(err));
  process.exit(1);
});

File: .skills/openclaw-skills/skills/kldsw/finance-daily-push/scripts/fetch-quote.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
腾讯财经 API 行情数据获取脚本
用于获取 A 股实时行情数据
"""

import requests
import sys
from datetime import datetime

def fetch_quotes(symbols):
    """
    获取腾讯财经 API 行情数据
    
    Args:
        symbols: 股票代码列表，如 ['sh000001', 'sz399006']
    
    Returns:
        dict: 行情数据字典
    """
    url = f'http://qt.gtimg.cn/q={",".join(symbols)}'
    
    headers = {
        'User-Agent': 'Mozilla/5.0',
        'Referer': 'https://finance.qq.com'
    }
    
    try:
        response = requests.get(url, headers=headers, timeout=10)
        
        if response.status_code != 200:
            print(f'数据获取失败，状态码：{response.status_code}', file=sys.stderr)
            return None
        
        content = response.content.decode('gbk')
        quotes = {}
        
        for line in content.strip().split('\n'):
            if '=' in line:
                # 解析：v_sh000001="51~上证指数~000001~4086.30~4049.91~..."
                data_part = line.split('"')[1] if '"' in line else line
                parts = data_part.split('~')
                
                if len(parts) > 10:
                    symbol = line.split('=')[0].replace('v_', '')
                    name = parts[1]
                    current = float(parts[3])
                    yesterday = float(parts[4])
                    change = current - yesterday
                    percent = (change / yesterday) * 100
                    
                    quotes[symbol] = {
                        'name': name,
                        'current': current,
                        'change': change,
                        'percent': percent,
                        'time': datetime.now().strftime('%H:%M')
                    }
        
        return quotes
        
    except Exception as e:
        print(f'获取行情数据异常：{e}', file=sys.stderr)
        return None


def main():
    """主函数"""
    # 默认获取主要指数
    default_symbols = [
        'sh000001',  # 上证指数
        'sz399006',  # 创业板指
        'sz399001',  # 深证成指
        'sh000016',  # 上证 50
        'sh000688',  # 科创 50
    ]
    
    # 从命令行参数获取股票代码
    if len(sys.argv) > 1:
        symbols = sys.argv[1].split(',')
    else:
        symbols = default_symbols
    
    quotes = fetch_quotes(symbols)
    
    if quotes:
        print(f'【来源：腾讯财经，{datetime.now().strftime("%H:%M")} 实时】')
        print('=' * 70)
        print('A 股主要指数')
        print('=' * 70)
        
        for symbol, data in quotes.items():
            print(f'{data["name"]:<10} {data["current"]:>10.2f}  '
                  f'涨跌：{data["change"]:+>8.2f} ({data["percent"]:+>7.2f}%)')
    else:
        print('数据获取失败', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()
